### PR TITLE
Avoid copies in `InlineTableScan` via TreeNode API

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?
Part of #9637  -- let's make DataFusion planning faster by not copying so much

## Rationale for this change

Now that we have the nice TreeNode API thanks to #8913 and @peter-toth  let's use it to both simplify the code and avoid copies

## What changes are included in this PR?

Rewrite `InlineTableScan` using TreeNode API

## Are these changes tested?
Existing CI

### CODE REVIEW

File: datafusion/optimizer/src/analyzer/inline_table_scan.rs [hunk 1]
```diff
@@ -51,65 +47,37 @@ impl AnalyzerRule for InlineTableScan {
...
-                .alias(table_name)?
-                .build()
-                .map(Transformed::yes)
-        }
-        LogicalPlan::Filter(filter) => {
-            let new_expr = filter.predicate.transform(&rewrite_subquery).data()?;
-            Filter::try_new(new_expr, filter.input)
-                .map(|e| Transformed::yes(LogicalPlan::Filter(e)))
+    // rewrite any subqueries in the plan first
+    let result = plan.map_subqueries(|plan| plan.transform_up(&analyze_internal))?;
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/inline_table_scan.rs:51]: This is the new API that @peter-toth  added in https://github.com/apache/arrow-datafusion/pull/9913

File: datafusion/optimizer/src/analyzer/inline_table_scan.rs [hunk 2]
```diff
@@ -51,65 +47,37 @@ impl AnalyzerRule for InlineTableScan {
...
+                    .map(Transformed::yes)
+            }
+            _ => Ok(Transformed::no(plan)),
         }
-        _ => Ok(Transformed::no(plan)),
-    }
-}
+    })?;
 
-fn rewrite_subquery(expr: Expr) -> Result<Transformed<Expr>> {
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/inline_table_scan.rs:85]: this is now handled by `apply_subqueries`

File: datafusion/optimizer/src/analyzer/inline_table_scan.rs [hunk 3]
```diff
@@ -51,65 +47,37 @@ impl AnalyzerRule for InlineTableScan {
...
-    match expr {
-        Expr::Exists(Exists { subquery, negated }) => {
-            let plan = subquery.subquery.as_ref().clone();
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/inline_table_scan.rs:88]: these clones are avoided in the current formulation

File: datafusion/optimizer/src/analyzer/inline_table_scan.rs (hunk 1 above)
Comment (crepererum) [inline datafusion/optimizer/src/analyzer/inline_table_scan.rs:51]:
```suggestion
    let plan = plan.map_subqueries(|plan| plan.transform_up(&analyze_internal))?;
```

When I read "result" I'm kinda expecting this to be a `Result<_, _>` which -- based on the `result.transform_data` call below -- is not the case here.

Same comment applies to the `let result` a few lines below.

**Review by crepererum — APPROVED** [review]
Small readability nitpick.

File: datafusion/optimizer/src/analyzer/inline_table_scan.rs (hunk 1 above)
Comment (alamb) [reply datafusion/optimizer/src/analyzer/inline_table_scan.rs]: It is a good call -- renamed to `transformed_plan` in 0e0ac8ee5

**Comment by alamb** [comment]
Thanks again @crepererum

**Pull request merged**